src/kaocha/type/clojure/spec/alpha/fdef.clj

```diff
@@ -0,0 +1,87 @@
+  (:require [clojure.spec.alpha :as s]
+            [clojure.spec.test.alpha :as stest]
+            [clojure.string :as str]
+            [clojure.test :as test]
+            [expound.alpha :as expound]
+            [kaocha.result :as result]
+            [kaocha.testable :as testable]
+            [kaocha.type :as type]))
+
+(alias 'stc 'clojure.spec.test.check)
+
+(defn stc-opt-key? [kw]
+  (some-> kw (namespace) (str/starts-with? "clojure.spec.test.check")))
+
+(defn stc-opt-keys [m]
+  (->> m (keys) (filter stc-opt-key?)))
+
+(defn stc-opts [m]
+  (->> m
+       (stc-opt-keys)
+       (select-keys m)))
+
+(defn load-testable [sym test-plan]
+  (let [nsname    (namespace sym)
+        test-name (str sym)
+        var       (resolve sym)]
+    {:kaocha.testable/type   :kaocha.type/clojure.spec.alpha.fdef
+     :kaocha.testable/id     (keyword test-name)
+     :kaocha.testable/meta   (meta var)
+     :kaocha.testable/desc   (str sym)
+     :kaocha.fdef/sym        sym
+     :kaocha.fdef/name       test-name
+     :kaocha.fdef/var        var
+     :kaocha.fdef/check-opts (stc-opts test-plan)}))
+
+(defn load-testables [syms]
+  (->> syms
+       (sort-by name)
+       (map load-testable)))
+
+(defn report-success [check-results]
+  (test/do-report
+   {:type    :pass
+    :message (str "Generative tests pass for "
+                  (str/join ", " (map :sym check-results)))}))
+
+(defn report-failure [check-results]
+  (doseq [failed-check (filter :failure check-results)]
+    (let [r       (stest/abbrev-result failed-check)
+          failure (:failure r)]
+      (test/do-report
+       {:type     :fail
+        :message  (expound/explain-results-str check-results)
+        :expected (->> r :spec rest (apply hash-map) :ret)
+        :actual   (if (instance? Throwable failure)
+                    failure
+                    (::stest/val failure))}))))
+
+(defmethod testable/-run :kaocha.type/clojure.spec.alpha.fdef
+  [{the-var  :kaocha.fdef/var
+    sym      :kaocha.fdef/sym
+    the-meta :kaocha.testable/meta
+    opts     :kaocha.fdef/check-opts
+    :as      testable} test-plan]
+  (type/with-report-counters
+    (let [check-results  (stest/check sym opts)
+          result-count   (count check-results)
+          checks-passed? (->> check-results (map :failure) (every? nil?))]
+      (if checks-passed?
+        (report-success check-results)
+        (report-failure check-results))
+      (merge testable {:kaocha.result/count result-count} (type/report-count)))))
+
+;; TODO: Why is this not defined in core?
+(s/def ::stc/opts (s/keys :opt [::stc/num-tests
+                                ::stc/max-size]))
+
+(s/def :kaocha.fdef/name :kaocha.var/name)
+(s/def :kaocha.fdef/var :kaocha.var/var)
+(s/def :kaocha.fdef/sym symbol?)
+
+(s/def :kaocha.type/clojure.spec.alpha.fdef
+  (s/keys :req [:kaocha.testable/type
+                :kaocha.testable/id
+                :kaocha.fdef/name
+                :kaocha.fdef/var]))
```

Comment:
There are a couple things missing here, like I said this part is the least well documented so that's my bad.

- `:kaocha.testable/wrap` this contains wrapping functions that need to be wrapped around your main test logic, so plugins like the output-capturing work
- `(hierarchy/derive! :kaocha.type/clojure.spec.alpha.fdef :kaocha.testable.type/leaf)` this test type never has children, so it's a leaf type. this is needed for some plugins and reporters.
- `(t/do-report ...)` I don't see any `:begin-...` and `:end-...` events for these test types. Does `stest/check` emit certain `clojure.test` events?

Reply:
I probably should add the `:begin-` and `:end-` reports, I haven't gotten it all running yet but I probably will today!